File: app/src/main/java/com/minibrowser/tab/TabManager.java
package com.minibrowser.tab;

import java.util.ArrayList;
import java.util.List;

public class TabManager {
    private static volatile TabManager instance;
    private final List<Tab> tabs = new ArrayList<>();
    private String currentUserId = "default_user";

    private TabManager() {}

    public static TabManager get() {
        if (instance == null) {
            synchronized (TabManager.class) {
                if (instance == null) instance = new TabManager();
            }
        }
        return instance;
    }

    public void setUserId(String userId) {
        this.currentUserId = userId != null ? userId : "default_user";
    }

    public synchronized void addTab(Tab tab) {
        if (tab != null && currentUserId.equals(tab.userId)) {
            tabs.add(tab);
        }
    }

    public synchronized List<Tab> getUserTabs() {
        List<Tab> userTabs = new ArrayList<>();
        for (Tab t : tabs) {
            if (currentUserId.equals(t.userId)) {
                userTabs.add(t);
            }
        }
        return userTabs;
    }

    public synchronized void closeTab(String id) {
        Tab toRemove = null;
        for (Tab t : tabs) {
            if (t.id.equals(id) && currentUserId.equals(t.userId)) {
                toRemove = t;
                break;
            }
        }
        if (toRemove != null) {
            tabs.remove(toRemove);
            toRemove.core.destroy();
        }
    }
}

File: app/src/main/java/com/minibrowser/tab/TabManager.kt
package com.minibrowser.tab

import java.util.ArrayList
import java.util.Collections

class TabManager private constructor() {
    private val tabs = Collections.synchronizedList(ArrayList<Tab>())
    private var currentUserId = "default_user"

    companion object {
        @JvmStatic
        val instance: TabManager by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            TabManager()
        }

        @JvmStatic
        fun get(): TabManager = instance
    }

    fun setUserId(userId: String?) {
        this.currentUserId = userId ?: "default_user"
    }

    fun addTab(tab: Tab?) {
        if (tab != null && currentUserId == tab.userId) {
            tabs.add(tab)
        }
    }

    fun getUserTabs(): List<Tab> {
        val userTabs = ArrayList<Tab>()
        synchronized(tabs) {
            for (t in tabs) {
                if (currentUserId == t.userId) {
                    userTabs.add(t)
                }
            }
        }
        return userTabs
    }

    fun closeTab(id: String) {
        var toRemove: Tab? = null
        synchronized(tabs) {
            for (t in tabs) {
                if (t.id == id && currentUserId == t.userId) {
                    toRemove = t
                    break
                }
            }
            if (toRemove != null) {
                tabs.remove(toRemove)
                toRemove?.core?.destroy()
            }
        }
    }
}
